zyte_common_items/pipelines.py

```diff
@@ -53,3 +53,82 @@ def __init__(self):
 
     def process_item(self, item, spider):
         return ae.downgrade(item)
+
+
+class DropLowProbabilityItemPipeline:
+    """This pipeline drops an item if its probability, defined in the settings,
+    is less than the specified threshold.
+
+    There are two ways that the thresholds can be defined per item type using
+    the ``ITEM_PROBABILITY_THRESHOLDS`` setting:
+
+    With paths to Item classes:
+        ITEM_PROBABILITY_THRESHOLDS = {
+            "zyte_common_items.items.Article": 0.1,
+            "zyte_common_items.items.Product": 0.2
+        }
+
+    With Item classes:
+        from zyte_common_items import Article, Product
+
+        ITEM_PROBABILITY_THRESHOLDS = {
+            Article: 0.1,
+            Product: 0.2
+        }
+    """
+
+    DEFAULT_ITEM_PROBABILITY_THRESHOLD = 0.1
+
+    def __init__(self, crawler):
+        self.stats = crawler.stats
+        self.thresholds_for_item = {}
+        self.default_threshold = None
+        self.init_thresholds(crawler.spider)
+
+    @classmethod
+    def from_crawler(cls, crawler):
+        return cls(crawler)
+
+    def init_thresholds(self, spider):
+        from scrapy.utils.misc import load_object
+
+        thresholds_settings = spider.settings.get("ITEM_PROBABILITY_THRESHOLDS", {})
```

Comment:
I wonder if we should make a copy here. Later thresholds_settings.pop is used, and it's not clear to me if it is problematic or not. It modifies settings in place, and I wonder if it breaks e.g. multiple spiders in the same process (they may share the same settings - Setting objects are going to be different, but the exact dict as a setting value could be shared - I'm not sure). It may also slightly break logging of settings ("default" will be missing), if for some reason it happens after init_thresholds method is executed.

Reply:
Oh, never thought that the settings can be share between spiders, but sure, let's make this much more reliable with deepcopy